TwoPageEBookReader/app/src/main/java/com/example/ebook_reader_sample/Book.kt

```diff
@@ -0,0 +1,202 @@
+                    }
+                }
+                bufferedReader.close()
+            }
+
+            buildPages()
+        }
+
+
+    private var currentParagraph = 0
+
+    private lateinit var pageStrings: ArrayList<ArrayList<String>>
+    fun getPageStrings(index: Int): ArrayList<String> {
+        return if (index < pageStrings.size && index >= 0) {
+            pageStrings[index]
+        }
+        else {
+            ArrayList()
+        }
+    }
+    val numPages
+        get() = pageStrings.size
+
+    var currentPage: Int
+        get(){
+            var page = 0
+            var remainingParagraphs = currentParagraph
+            while (page < pageStrings.size && pageStrings[page].size <= remainingParagraphs ) {
+                remainingParagraphs -= pageStrings[page].size
+                page += 1
+            }
+            return page
+        }
+        set(inPage) {
+            currentParagraph = 0
+            for (page in 0 until min(inPage, pageStrings.size)) {
+                currentParagraph += pageStrings[page].size
+            }
+        }
+
+    val pagePadding = activityContext.resources.getDimension(R.dimen.page_padding).toInt()
+    var pageRects = ArrayList<Rect>()
+        set(input) {
+            field = input
+            buildPages()
+        }
+
+    var fontSize = 16
+        set(input) {
+            field = input
+            buildPages()
+        }
+
+    //buildPages() uses the available rects of the current layout to organize the chapter's strings into pages
+    //TODO buildPages() supports multiple pages by cycling through pageRects instead of using only one Rect of constraints
+    fun buildPages() {
+        if (pageRects.isEmpty()) { return }
+
+        pageStrings = ArrayList()
+        var textStrings = ArrayList<String>()
+
+        var pageRectIndex = 0
+        var availableHeight = pageRects[pageRectIndex].height() - (2 * pagePadding)
+        var widthSpec = View.MeasureSpec.makeMeasureSpec(pageRects[pageRectIndex].width() - (2 * pagePadding), View.MeasureSpec.AT_MOST)
+        val heightSpec = View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED)
+        val paragraphStringsCopy = paragraphStrings.clone() as ArrayList<String>
+        val paragraphIt = paragraphStringsCopy.listIterator()
+
+        val textView = View.inflate(activityContext, R.layout.book_page_text_paragraph, null) as TextView
+        textView.layoutParams = LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT)
+        textView.textSize = fontSize.toFloat()
+        while (paragraphIt.hasNext()) {
+            val paragraphString = paragraphIt.next()
+
+            textView.text = Html.fromHtml(paragraphString)
+            textView.measure(widthSpec,heightSpec)
+            if (textView.measuredHeight > availableHeight) {
+                var splitIndex = paragraphString.lastIndexOf(". ")
```

Comment:
probably you can move the `if` body to its own function. The current function is too large and difficult to read, if you split the body to small functions you can read way better what this does. e.g:
```
.....
doSomething()
doNowThis()
endWithThis()
```

Reply:
excising parts would involve passing around a lot of variables

I made an effort to shorten the function - hopefully this improves readability enough